### PR TITLE
Ada C22 Phoenix Class Madina Dzhetegenova

### PR DESCRIPTION
**Pull request by Madina-j**

### CODE REVIEW

File: src/components/ChatEntry.jsx [hunk 1]
```diff
@@ -1,20 +1,32 @@
 import './ChatEntry.css';
+import TimeStamp from './TimeStamp';
+import PropTypes from 'prop-types';
+// import { useState } from 'react';
 
-const ChatEntry = () => {
+const ChatEntry = (props) => {
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:6]: `ChatEntry` uses a single `props` param, but `ChatLog` uses a destructured object. Personally, I prefer the destructured style, since it makes the expected component attributes a bit more clear. And it's fine to use a mixture of styles in this project, but in general, try to pick one style or the other.

File: src/components/ChatEntry.jsx [hunk 2]
```diff
@@ -1,20 +1,32 @@
...
   return (
     <div className="chat-entry local">
-      <h2 className="entry-name">Replace with name of sender</h2>
+      <h2 className="entry-name">{props.sender}</h2>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{props.body}</p>
+        <p className="entry-time"><TimeStamp time={props.timeStamp}/></p>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:16]: Nice use of the supplied `TimeStamp`. All we need to do is pass in the `timeStamp` string from the message data and it takes care of the rest. All we had to do was confirm the name and type of the prop it was expecting (which we could do through its PropTypes) and we're all set!

File: src/components/ChatEntry.jsx [hunk 3]
```diff
@@ -1,20 +1,32 @@
...
 };
 
 ChatEntry.propTypes = {
-  // Fill with correct proptypes
+  id: PropTypes.number.isRequired,
+  sender: PropTypes.string.isRequired,
+  body: PropTypes.string.isRequired,
+  timeStamp: PropTypes.string.isRequired,
+  liked: PropTypes.bool,
+  onLikeToggle: PropTypes.func.isRequired,
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:29]: 👀 The `id`, `sender`, `body`, `timeStamp`, and `liked` props are always passed (they're defined explicitly in the data and also provided in the test) so we can (and should) mark them `isRequired`. 

The remaining props were up to you, and the tests don't know about them. As a result, using `isRequired` causes a warning when running any tests that only pass the known props. If you didn't see those warnings when running the tests, be sure to also try running the terminal `npm test` since the warnings are more visible.

To properly mark any other props `isRequired`, we would also need to update the tests to include at least dummy values (such as an empty callback `() => {}` for the like handler) to make the proptype checking happy.

Alternatively, for any props that we leave not required, we should also have logic in our component to not try to use the value if it's undefined.

File: src/components/ChatLog.jsx [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  entries: PropTypes.arrayOf(
+    PropTypes.shape({
+      id: PropTypes.number.isRequired,
+      sender: PropTypes.string.isRequired,
+      body: PropTypes.string.isRequired,
+      timeStamp: PropTypes.string.isRequired,
+      liked: PropTypes.bool,
+    })
+  ).isRequired,
+  onLikeToggle: PropTypes.func.isRequired
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:33]: Similar to the props for `ChatEntry` here, the `entries` prop is included in the tests, but the like toggle is not, resulting in prop warnings (unless we update the tests to reflect our custom props).

Again, if we were to leave this as not required so as to avoid the test warnings, we'd want to be sure that all the script logic in our component worked properly even in the absence of this value.

File: src/components/ChatLog.jsx [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+};
+
+ChatLog.propTypes = {
+  entries: PropTypes.arrayOf(
+    PropTypes.shape({
+      id: PropTypes.number.isRequired,
+      sender: PropTypes.string.isRequired,
+      body: PropTypes.string.isRequired,
+      timeStamp: PropTypes.string.isRequired,
+      liked: PropTypes.bool,
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:30]: We can make the `liked` value required, since even in the base tests a value is being supplied.

File: src/components/ChatLog.jsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import PropTypes from 'prop-types';
+import ChatEntry from './ChatEntry';
+import './ChatLog.css';
+
+const ChatLog = ({entries, onLikeToggle}) => {
+  const chatComponents = entries.map((entry) => {
+    return (
+      <ChatEntry
+        key={entry.id}
+        id={entry.id}
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:10]: 👍 The `key` attribute is important for React to be able to detect certain kinds of data changes in an efficient manner. We're also using the `id` for our own `id` prop, so it might feel redundant to pass both, but one is for our logic and one is for React internals (we can't safely access the `key` value in any meaningful way).

File: src/components/ChatLog.jsx [hunk 7]
```diff
@@ -0,0 +1,36 @@
+import PropTypes from 'prop-types';
+import ChatEntry from './ChatEntry';
+import './ChatLog.css';
+
+const ChatLog = ({entries, onLikeToggle}) => {
+  const chatComponents = entries.map((entry) => {
```
Comment (anselrognlie) [inline src/components/ChatLog.jsx:6]: Nice use of `map` to convert from the message data into `ChatEntry` components. We can perform this mapping storing the result into a variable we use in the JSX result as you did here (components are functions, so we can run JS code as usual before we reach the `return`, and even sometimes have multiple `return` statements with different JSX), we could make a helper function that we call as part of the `return`, or this expression itself could be part of the `return` JSX, which I often like since it helps me see the overall structure of the component, though can make debugging a little more tricky. But any of those approaches will work fine.

File: src/App.jsx [hunk 8]
```diff
@@ -1,14 +1,30 @@
 import './App.css';
+import ChatLog from './components/ChatLog';
+import chatData from './data/messages.json';
+import { useState } from 'react';
 
 const App = () => {
+  const [entries, setEntries] = useState(chatData);
+  const toggleLike = (id) => {
```
Comment (anselrognlie) [inline src/App.jsx:8]: 👍 Since our state is defined here, we also need to define our mutating function here so that it can "see" the setter function. All we need to receive is the `id` of the message to toggle, which allows us to locate the message to update, and calculate the next state value. We can pass this all the way down to our `ChatEntry` which handles the click event, passing us the `id` of the message that was clicked.

File: src/App.jsx [hunk 9]
```diff
@@ -1,14 +1,30 @@
...
+    const updatedEntries = entries.map((entry) => {
+      if (entry.id === id) {
+        return { ...entry, liked: !entry.liked };
+      } else {
+        return entry;
+      }
+    });
+    setEntries(updatedEntries);
```
Comment (anselrognlie) [inline src/App.jsx:16]: In this case, calculating the next version of the message data using the current state variable and passing that updated version to the state setter shouldn't cause any problems, but we still generally prefer using the callback style of setters. Using that approach, we pass a function to the setter whose parameter will receive the latest state value, and which returns the new value to use for the state.
```js
    setEntries(entries => entries.map(entry => {
      if (entry.id === id) {
        return {...entry, liked: !entry.liked};
      } else {
        return entry;
      };
    }));
```

File: src/App.jsx [hunk 10]
```diff
@@ -1,14 +1,30 @@
 import './App.css';
+import ChatLog from './components/ChatLog';
+import chatData from './data/messages.json';
+import { useState } from 'react';
 
 const App = () => {
+  const [entries, setEntries] = useState(chatData);
+  const toggleLike = (id) => {
+    const updatedEntries = entries.map((entry) => {
```
Comment (anselrognlie) [inline src/App.jsx:9]: Nice use of map here to both handle making a new list so that React sees the message data has changed, and make new data for the clicked message with its like status toggled.

File: src/App.jsx [hunk 11]
```diff
@@ -1,14 +1,30 @@
...
+        return { ...entry, liked: !entry.liked };
+      } else {
+        return entry;
+      }
+    });
+    setEntries(updatedEntries);
+  };
+  const likeCount = entries.reduce((total, entry) => {
+    return entry.liked ? total + 1 : total;
+  }, 0);
```
Comment (anselrognlie) [inline src/App.jsx:20]: Nice job determining the total likes based on the like data of each message. We don't need an additional piece of state to track this, since it can be derived from the existing state we are tracking.

One way that we can make using `reduce` a little more understandable for other programmers reading our code would be to give the custom function that we pass to `reduce` a descriptive name.
```js
const sumLikes = (total, entry) => {
  return entry.liked ? total + 1 : total;
};

const likeCount = entries.reduce(sumLikes, 0);

File: src/components/ChatEntry.jsx [hunk 12]
```diff
@@ -1,20 +1,32 @@
 import './ChatEntry.css';
+import TimeStamp from './TimeStamp';
+import PropTypes from 'prop-types';
+// import { useState } from 'react';
 
-const ChatEntry = () => {
+const ChatEntry = (props) => {
+  const heart = props.liked ? '❤️' : '🤍';
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:7]: 👍 We can figure out which emoji to use for the liked status based on the `liked` prop without creating any additional state.

File: src/components/ChatEntry.jsx [hunk 13]
```diff
@@ -1,20 +1,32 @@
...
+  const handleLikeClick = () => {
+    props.onLikeToggle(props.id);
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:9]: 👍 Passing the `id` of this message lets the logic defined up in the `App` find the message to update in its data.

File: src/components/ChatEntry.jsx [hunk 14]
```diff
@@ -1,20 +1,32 @@
...
     <div className="chat-entry local">
-      <h2 className="entry-name">Replace with name of sender</h2>
+      <h2 className="entry-name">{props.sender}</h2>
       <section className="entry-bubble">
-        <p>Replace with body of ChatEntry</p>
-        <p className="entry-time">Replace with TimeStamp component</p>
-        <button className="like">🤍</button>
+        <p>{props.body}</p>
+        <p className="entry-time"><TimeStamp time={props.timeStamp}/></p>
+        <button className="like" onClick={handleLikeClick}>{heart}</button>
```
Comment (anselrognlie) [inline src/components/ChatEntry.jsx:17]: 👍 We need a wrapper of some kind rather than calling the received callback through props, since our callback function is expecting a message id as its parameter. If we tried to use it directly as the click event handler, React would end up passing it a clink event, since any function registered as an event handler will always be given the event detail information as its argument.

File: src/App.jsx [hunk 15]
```diff
@@ -1,14 +1,30 @@
...
   return (
     <div id="App">
       <header>
-        <h1>Application title</h1>
+        <h1>{likeCount} ❤️s</h1>
```
Comment (anselrognlie) [inline src/App.jsx:24]: 👀 Notice in the example screenshots that in addition to the like counts, there was a header listing the conversation participants that we wanted to show, even if it weren't calculated dynamically.

**Review by anselrognlie — COMMENTED** [review]
Looks good! Please review my comments, and let me know if you have any questions. Nice job!
